aten/src/ATen/native/Distributions.cpp

```diff
@@ -487,6 +487,17 @@ Tensor& multinomial_out(Tensor& result, const Tensor& self, int64_t n_sample, bo
   } else {
     result.resize_({n_sample});
   }
+
+  // Fast-path based on RobertoLat example.
+  // Reference:
+  // https://github.com/pytorch/pytorch/issues/11931#issuecomment-625882503
+  // Half is not supported on CPU.
+  if (!with_replacement &&
+      !(self.device().is_cpu() && self.scalar_type() == ScalarType::Half)) {
+    auto rand = at::rand_like(self);
```

Comment:
So in all, to fix both result in-place update and `gen` argument you'd need
```suggestion
    auto rand = at::empty_like(self).uniform_(0, 1, gen);
    rand.log_().div_(self); //save memory with inplace operations
    Tensor vals;
    at::topk_out(vals, result, rand, n_sample);
    return result;
```